src/nodecache.h

```diff
@@ -13,55 +13,50 @@
 	every time a new tree is generated.
 */
 
-typedef std::pair<long, long> IndexPair;
-
 class NodeCache
 {
 public:
     NodeCache() { }
-    virtual ~NodeCache() { cache.clear(); }
+    virtual ~NodeCache() { }
 
     bool contains(const AbstractNode &node) const {
-        auto result = cache.find(node.index()); 
-        return result != cache.end() && result->second.second >= 0L;
+        auto result = this->cache.find(node.index()); 
+        return result != this->cache.end() && 
+            result->second.second >= 0L && 
+            this->rootString.size() >= result->second.second;
     }
 
-    const std::string operator[](const AbstractNode &node) const {
-        auto result = cache.find(node.index());
-        assert(result != cache.end() && "nodecache miss");
-        auto start = result->second.first;
-        auto end = result->second.second;
-        return root_string.substr(start, end-start);
+    std::string operator[](const AbstractNode &node) const {
+        // throws std::out_of_range on miss
+        auto indexpair = this->cache.at(node.index()); 
+        return rootString.substr(indexpair.first, indexpair.second - indexpair.first);
     }
 
     void insert_start(const size_t nodeidx, const long startindex) {
-        assert(cache.find(nodeidx) == cache.end() && "start index inserted twice");
-        cache[nodeidx] = std::make_pair(startindex, -1L);
+        assert(this->cache.count(nodeidx) == 0 && "start index inserted twice");
+        this->cache.emplace(nodeidx, std::make_pair(startindex, -1L));
     }
 
     void insert_end(const size_t nodeidx, const long endindex) {
-        auto result = cache.find(nodeidx);
-        assert(result != cache.end() && "end index inserted before start");
-        auto indexpair = result->second;
+        // throws std::out_of_range on miss
+        auto indexpair = this->cache.at(nodeidx); 
         assert(indexpair.second == -1L && "end index inserted twice");
-        cache[nodeidx] = std::make_pair(indexpair.first, endindex);
+        this->cache[nodeidx] = std::make_pair(indexpair.first, endindex);
```

Comment:
Since you opted for `emplace()` in the `insert_start()` method, we should probably `emplace()` here as well

Reply:
As I understand it, emplace is for adding a key where there isn't one yet.  

From http://en.cppreference.com/w/cpp/container/map/emplace
>Inserts a new element into the container constructed in-place with the given args if there is no element with the key in the container.

When we insert the end index, we are updating the value(second part of the pair) for an existing key, so its a little different.

Reply:
You’re right